### PR TITLE
Add missing ranked stats

### PR DESCRIPTION
**Pull request by byBlurr**
Added ranked TimeUpdated, Kills, Deaths, NextRankMMR, PreviousRankMMR, SkillMean, SkillStdev, LastMatchResult, LastMatchMMRChange, LastMatchSkillChange, LastMatchSkillStdevChange.

SkillStdev and LastMatchSkillStdevChange possibly need renaming as I am not too sure on what "stdev" stands for.

Also, feel free to leave me with some 'todos' towards the repo.

### CODE REVIEW

**Comment by aspriddell** [comment]
`stdev` is also known as standard deviation. It is basically an "uncertainty" factor against the ranking to show how close the average ranking values are together. The lower the value, the more confident the ranking system is that they are in the right band.

I'll have a quick look at your edits and act from there.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs [hunk 1]
```diff
@@ -39,6 +48,31 @@ public class Season
         [JsonProperty("maxmmr")]
         public double MaxMMR { get; set; }
 
+        [JsonProperty("next_rank_mmr")]
```
Comment (aspriddell) [inline DragonFruit.Six.API/Stats/SeasonalStats.cs]: These property headers aren't used on ubi's side so if you want to name them something easier to understand then feel free. These will probably be used in a backend if it were the case.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs [hunk 2]
```diff
@@ -39,6 +48,31 @@ public class Season
...
+
+        [JsonProperty("last_match_mmr_change")]
+        public double LastMatchMMRChange { get; set; }
+
+        [JsonProperty("last_match_skill_mean_change")]
+        public double LastMatchSkillChange { get; set; }
+
+        [JsonProperty("last_match_skill_stdev_change")]
+        public double LastMatchSkillStdevChange { get; set; } // TODO: Wasn't sure what to call this... Needs to be named more appropriately
+
```
Comment (aspriddell) [inline DragonFruit.Six.API/Stats/SeasonalStats.cs]: Add `top_rank_position`? It equals `0` on a user that isn't in the top 10k so a `uint` will suffice here

File: DragonFruit.Six.API/Stats/SeasonalStats.cs [hunk 3]
```diff
@@ -15,9 +15,18 @@ public class Season
         [JsonProperty("guid")]
         public string Guid { get; set; }
 
+        [JsonProperty("update_time")]
+        public string TimeUpdated { get; set; }
```
[the remark this answers is not in the document]
Comment (aspriddell) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: Also on this, the backend returns `1970-01-01T00:00:00+00:00` for players that haven't played any ranked matches in that season. Just a suggestion but if we said any dates posted before the release date of R6 is counted as not played (not that it has to go in the api)?

File: DragonFruit.Six.API/Stats/SeasonalStats.cs [hunk 4]
```diff
@@ -39,6 +48,31 @@ public class Season
...
+        public double PreviousRankMMR { get; set; }
+
+        [JsonProperty("skill_mean")]
+        public double SkillMean { get; set; }
+
+        [JsonProperty("skill_stdev")]
+        public double SkillStdev { get; set; } // TODO: Wasn't sure what to call this... Needs to be named more appropriately
+
+        [JsonProperty("last_match_result")]
+        public uint LastMatchResult { get; set; }
```
[the remark this answers is not in the document]
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: I think they are win, loss and draw. Hard to confirm the draw. I can test this today.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs [hunk 5]
```diff
@@ -39,6 +48,31 @@ public class Season
...
+        public double NextRankMMR { get; set; }
+
+        [JsonProperty("previous_rank_mmr")]
+        public double PreviousRankMMR { get; set; }
+
+        [JsonProperty("skill_mean")]
+        public double SkillMean { get; set; }
+
+        [JsonProperty("skill_stdev")]
+        public double SkillStdev { get; set; } // TODO: Wasn't sure what to call this... Needs to be named more appropriately
```
[the remark this answers is not in the document]
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: `SkillUncertainty` I think would be the better option, I will make this change today.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: I don't think this is needed within the api personally.

[the remark this answers is not in the document]
Comment (aspriddell) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: Agreed. Taking the suggested change would be a good idea for use in calculations or whatever

**Comment by aspriddell** [comment]
@byBlurr re-request a review once your happy and I'll take a look

File: DragonFruit.Six.API/Stats/SeasonalStats.cs (hunk 1 above)
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: I think it is good to keep the headers consistent with Ubisoft's side. I will let you decide however.

Comment (aspriddell) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: we can leave these as they are. They aren't as bad as some of the stats ones, like the general ones.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (aspriddell) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: Is it even possible to draw on a ranked match? I would assume Win = 1 and loss = 0

[the remark this answers is not in the document]
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: You are completely correct, it is not possible to draw on ranked. I had a bit of a brain fart there as it is possible to draw on custom match.

An `enum` works here.

**Comment by byBlurr** [comment]
I am currently getting Error 502 (Bad Gateway). I will resume testing shortly.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs (hunk 2 above)
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: I have added this in my next commit.

**Comment by byBlurr** [comment]
Error 502 (Bad Gateway) is still a problem, every so often it will connect for a few minutes. All stat websites are having issues loading recent matches and I am having issues even loading the main menu of the game. So we will have to play the waiting game for now.

File: DragonFruit.Six.API/Stats/SeasonalStats.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (byBlurr) [reply DragonFruit.Six.API/Stats/SeasonalStats.cs]: When entering a Uplay name that doesn't have any playtime, the match result returns 0. On my Uplay account it returns 2, although I do not know if that is a win or loss, seems I have played two games today and neither of them show up in my stats anywhere (my MMR has not even updated in game since my last game).

The `enum` will have to wait until we can check the result for win or loss. I believe default is 0, loss 1 and win 2.

**Comment by aspriddell** [comment]
Your build is now passing, and I've updated the branch to even with #47

**Comment by byBlurr** [comment]
Added KD and WL to seasonal to make it more consistent with general stats.

**Comment by aspriddell** [comment]
For the last match result enum I have learned some things:

0 = unknown/not played (as you said)
1 = win (not checked though safe to assume)
2 = loss
3 = abandon (the one that we didn't know about)

File: DragonFruit.Six.API/Processing/RawConvert.cs [hunk 6]
```diff
@@ -118,15 +118,30 @@ public static Season ToSeason(this JObject jObject, string guid)
...
                 Guid = guid,
-
+                TimeUpdated = DateTime.Parse(json.GetString(SeasonalRanked.TimeUpdated, DateTime.Now.ToString())),
                 SeasonId = json.GetByte(SeasonalRanked.Season),
-                Rank = json.GetUInt(SeasonalRanked.Rank),
-                MaxRank = json.GetUInt(SeasonalRanked.MaxRank),
+
+                Kills = json.GetUInt(SeasonalRanked.Kills),
+                Deaths = json.GetUInt(SeasonalRanked.Deaths),
+                KD = (json.GetFloat(SeasonalRanked.Kills) / json.GetFloat(SeasonalRanked.Deaths)),
```
Comment (aspriddell) [inline DragonFruit.Six.API/Processing/RawConvert.cs]:
```suggestion
                KD = (json.GetFloat(SeasonalRanked.Kills, 1) / json.GetFloat(SeasonalRanked.Deaths, 1)),
```
They need to default to 1 if they are in a KD or WL calculation as otherwise we get a `NaN` result

File: DragonFruit.Six.API/Processing/RawConvert.cs [hunk 7]
```diff
@@ -118,15 +118,30 @@ public static Season ToSeason(this JObject jObject, string guid)
...
                 Wins = json.GetUInt(SeasonalRanked.Wins),
                 Losses = json.GetUInt(SeasonalRanked.Losses),
+                WL = (json.GetFloat(SeasonalRanked.Wins) / json.GetFloat(SeasonalRanked.Losses)),
```
Comment (aspriddell) [inline DragonFruit.Six.API/Processing/RawConvert.cs]: same as above

Comment (aspriddell) [inline DragonFruit.Six.API/Processing/RawConvert.cs]:
```suggestion
                WL = (json.GetFloat(SeasonalRanked.Wins, 1) / json.GetFloat(SeasonalRanked.Losses, 1)),
```

**Comment by aspriddell** [comment]
will merge once the build passes
